File: 1776-minimum-operations-to-reduce-x-to-zero/minimum-operations-to-reduce-x-to-zero.cpp
class Solution {
public:
    int minOperations(vector<int>& a, int x) {
        int n = a.size();
        vector<int>suffix(n+1, 0);

        map<int,int> suffixSumIdx;
        for(int i = n-1; i >= 0; i--) {
            suffix[i] = suffix[i+1] + a[i];
            if(suffixSumIdx.count(suffix[i]) == 0) {
                suffixSumIdx[suffix[i]] = i;
            }
        }

        int prefixSum = 0;
        int res = INT_MAX;
        if(suffixSumIdx.count(x)) {
            res = n - suffixSumIdx[x];
        }

        for(int i =0 ; i < n ;i++) {
            prefixSum += a[i];
            if(prefixSum == x) {
                res = min(res, i+1);
            }

            int remainingSum = x - prefixSum;
            if(suffixSumIdx.count(remainingSum)) {
                int j = suffixSumIdx[remainingSum];
                if(j <= i) continue;

                int front = i+1;
                int back = n - j;
                res = min(res, front + back);
            }
        }
        
        if(res == INT_MAX) return -1;
        return res;
    }
};

File: 1776-minimum-operations-to-reduce-x-to-zero/minimum-operations-to-reduce-x-to-zero.java
class Solution {
    public int minOperations(int[] a, int x) {
        int n = a.length;
        
        int[] suffix = new int[n+1];

        Map<Integer, Integer> suffixSumIdx = new HashMap<>();
        for(int i = n-1; i >= 0; i--) {
            suffix[i] = suffix[i+1] + a[i];
            if(!suffixSumIdx.containsKey(suffix[i])) {
                suffixSumIdx.put(suffix[i],i);
            }
        }

        int prefixSum = 0;
        int res = Integer.MAX_VALUE;
        if(suffixSumIdx.containsKey(x)) {
            res = n - suffixSumIdx.get(x);
        }

        for(int i =0 ; i < n ;i++) {
            prefixSum += a[i];
            if(prefixSum == x) {
                res = Math.min(res, i+1);
            }

            int remainingSum = x - prefixSum;
            if(suffixSumIdx.containsKey(remainingSum)) {
                int j = suffixSumIdx.get(remainingSum);
                if(j <= i) continue;

                int front = i+1;
                int back = n - j;
                res = Math.min(res, front + back);
            }
        }
        
        if(res == Integer.MAX_VALUE) return -1;
        return res;
        
    }
}